### DISCUSSION TITLE
761

### DISCUSSION THREAD

**Discussion by giscus** [jojoldu/blog-comments#928] [Comments]
# 761

얼마 전에 2개의 핫한 컨텐츠가 공유되었다. 존경하는 재민님의 유튜브 - 테스트에서 @Transactional 을 사용해야 할까? 존경하는 토비님의 페이스북 2개의 컨텐츠에서 테스트 데이터 초기화에 @Transactional 사용하는 것에 대해 서로 다른 의견을 내신 것이다. 마침 페이스북에 태깅되기도 했고 (ㅠㅠ) 과거에 라이브 방송에서도 "향로님은 반대한다" 라고 언급되기도 했었다. (반대하는 것은 사실이기도 하고..) 내 생각을 정리해야지 해야지 하다가, 마침 이번주에 시간이 되어서 정리하게 되었다. 1. Spring Team은? 내 의견을 정리하기 전에, 먼저 Spring Team 의 코드를 살펴보자. 인프라스트럭쳐 계층 (데이터베이스) 테스트를 작성하는 팀의 코드를 보면 될 것 같아서 Sprin..

https://jojoldu.tistory.com/761

**Comment by YuAnsang** [comment]
안녕하세요.

마지막 "테스트 데이터 초기화" 부분에 대해서 질문이 좀 있습니다!

저도 테스트 코드 작성 시 데이터가 필요하면 대부분 데이터 초기화 & truncate table로 작성을 하는데요.

이게 테스트코드의 양이 늘어나면 해당 부분 때문에 생각보다 테스트가 오래돌더라구요...

향로님이 경험하신 조직들의 테스트코드가 저희 조직보다는 훨씬 테스트코드도 많을 것 같은데, 혹시 관련해서는 어떻게 처리하시나요? 

항상 좋은글 남겨주셔서 감사드리고, 많이 배워 갑니다 :)

**Comment by havefunatcode** [comment]
안녕하세요. 항상 좋은 글 잘 보고 있습니다.
여쭤보고 싶은 부분이 있어서 댓글 남깁니다.
'모든 테이블간에 FK 제약 조건을 사용하지 않는다.' 라는 제약조건에 질문이 있습니다.
RDB를 사용하는 이유 중 가장 큰 이유가 테이블 간의 관계라고 생각하는데, FK 제약 조건을 사용하지 않는다면 이러한 이유가 많이 희석된다고 생각합니다. 저의 짧은 식견으로는 FK를 사용하지 않으면서 RDB의 장점을 살리는 방법이 생각나지 않습니다.

향로님께서는 FK를 사용하지 않는다 라는 제약 조건을 지키면서 RDB의 장점을 살리는 방법이 있으신지, 있으시다면 어떠한 방식으로 테이블 설계를 구성하시는지 궁금합니다.

**Reply by havefunatcode** [reply]
Github에서 FK 제약 조건을 사용하지 않는다는 아티클이 있어서 공유드립니다.
https://github.com/github/gh-ost/issues/331

**Comment by sinbom** [comment]
안녕하세요. 좋은 글 잘 읽었습니다.

저도 같은 생각을 가지고 가급적 테스트 코드에서 @Transacntional을 사용하지 않고 있는데요.
다만 @DataJpaTest와 같은 영속성 레이어에 대한 단위 테스트를 할 때는 사용하고 있기도 합니다. (@DataJpaTest에 포함되어 있음)
그 이유는 위 글에서 말씀하신 상황들은 통합 테스트를 하게되었을 때 발생할 수 있다고 생각하기 때문입니다.
혹시 이 부분에 대해서는 어떤 생각을 가지고 계신지 피드백 주시면 정말 감사드리겠습니다!

**Comment by highway92** [comment]
요약 :  S3저장소에 대한 테스트는 어떻게 진행하는게 좋을까요?

주제와는 조금 거리가 있지만 질문남깁니다.

DB와 관련된 테스트를 할때야 인메모리 DB나 Transaction을 사용하여 처리할 수 있지만
다른 외부 저장소 의존성을 가진 경우에는 어떻게 테스트를 진행하시는지 궁금합니다.

제가 겪고 있는 상황은 S3저장소와 관련된 것인데
Repository Interface를 구현한 mocking으로 테스트를 진행하자니 테스트에 의미가 없는것 같아서 질문 남깁니다.

결과적으로는 TEST를 위한 Bucket을 하나 만들어서 진행할 예정이긴 합니다

**Comment by jobmania** [comment]
좋은내용으로 많이 배우고 갑니다! 항상 감사합니다 : ]
